streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/TaskAssignor.java

```diff
@@ -0,0 +1,195 @@
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+
+public class TaskAssignor<C, T extends Comparable<T>> {
+
+    private static final Logger log = LoggerFactory.getLogger(TaskAssignor.class);
+
+    public static <C, T extends Comparable<T>> Map<C, ClientState<T>> assign(Map<C, ClientState<T>> states, Set<T> tasks, int numStandbyReplicas) {
+        long seed = 0L;
+        for (C client : states.keySet()) {
+            seed += client.hashCode();
+        }
+
+        TaskAssignor<C, T> assignor = new TaskAssignor<>(states, tasks, seed);
+        assignor.assignTasks();
+        if (numStandbyReplicas > 0)
+            assignor.assignStandbyTasks(numStandbyReplicas);
+
+        return assignor.states;
+    }
+
+    private final Random rand;
+    private final Map<C, ClientState<T>> states;
+    private final Set<TaskPair<T>> taskPairs;
+    private final int maxNumTaskPairs;
+    private final ArrayList<T> tasks;
+
+    private TaskAssignor(Map<C, ClientState<T>> states, Set<T> tasks, long randomSeed) {
+        this.rand = new Random(randomSeed);
+        this.states = new HashMap<>();
+        for (Map.Entry<C, ClientState<T>> entry : states.entrySet()) {
+            this.states.put(entry.getKey(), entry.getValue().copy());
+        }
+        this.tasks = new ArrayList<>(tasks);
+
+        int numTasks = tasks.size();
+        this.maxNumTaskPairs = numTasks * (numTasks - 1) / 2;
+        this.taskPairs = new HashSet<>(this.maxNumTaskPairs);
+    }
+
+    public void assignTasks() {
+        assignTasks(true);
+    }
+
+    public void assignStandbyTasks(int numStandbyReplicas) {
+        int numReplicas = Math.min(numStandbyReplicas, states.size() - 1);
+        for (int i = 0; i < numReplicas; i++) {
+            assignTasks(false);
+        }
+    }
+
+    private void assignTasks(boolean active) {
+        Collections.shuffle(this.tasks, rand);
+
+        for (T task : tasks) {
+            ClientState<T> state = findClientFor(task);
+
+            if (state != null) {
+                state.assign(task, active);
+            } else {
+                TaskAssignmentException ex = new TaskAssignmentException("failed to find an assignable client");
+                log.error(ex.getMessage(), ex);
+                throw ex;
+            }
+        }
+    }
+
+    private ClientState<T> findClientFor(T task) {
+        boolean checkTaskPairs = taskPairs.size() < maxNumTaskPairs;
+
+        ClientState<T> state = findClientByAdditionCost(task, checkTaskPairs);
+
+        if (state == null && checkTaskPairs)
+            state = findClientByAdditionCost(task, false);
+
+        if (state != null)
+            addTaskPairs(task, state);
+
+        return state;
+    }
+
+    private ClientState<T> findClientByAdditionCost(T task, boolean checkTaskPairs) {
```

Comment:
With this logic we can still possibly assign a task to another client that has not been loaded yet instead of the client that it is previously assigned to, right? If yes we need to update the explanation comments.